File: neet-code/backtracking/combination-sum.py
# https://leetcode.com/problems/combination-sum/

class Solution(object):
    def combinationSum(self, candidates, target):
        n = len(candidates)
        result = []

        def backtrack(remain, comb, start):
            if remain == 0:
                result.append(comb[:])
                return
            if remain < 0:
                return 

            for index in range(start, n):
                comb.append(candidates[index])
                backtrack(remain - candidates[index], comb, index)
                comb.pop()
            
        backtrack(target, [], 0)
        return result
        

File: data-structures/backtracking/combinationSum.js
function combinationSum(nums, target) {
  const result = [];

  function dfs(index, current, total) {
    if (total === target) {
      result.push([...current]);
      return;
    }

    if (index >= nums.length || total > target) return;

    current.push(nums[index]);
    dfs(index, current, total + nums[index]);

    current.pop();
    dfs(index + 1, current, total);
  }

  dfs(0, [], 0);
  return result;
}
